x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java

```diff
@@ -0,0 +1,391 @@
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.io.stream.ByteArrayStreamInput;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.index.fielddata.AbstractSortedSetDocValues;
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.FieldDataContext;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.LeafOrdinalsFieldData;
+import org.elasticsearch.index.fielddata.plain.AbstractIndexOrdinalsFieldData;
+import org.elasticsearch.index.fielddata.plain.AbstractLeafOrdinalsFieldData;
+import org.elasticsearch.index.mapper.BinaryFieldMapper;
+import org.elasticsearch.index.mapper.DocumentParserContext;
+import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.KeywordFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.Mapper;
+import org.elasticsearch.index.mapper.MapperBuilderContext;
+import org.elasticsearch.index.mapper.SourceValueFetcher;
+import org.elasticsearch.index.mapper.StringFieldType;
+import org.elasticsearch.index.mapper.TextSearchInfo;
+import org.elasticsearch.index.mapper.ValueFetcher;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.script.field.KeywordDocValuesField;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.MultiValueMode;
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+import org.elasticsearch.search.sort.BucketedSort;
+import org.elasticsearch.search.sort.SortOrder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import static org.elasticsearch.common.lucene.Lucene.KEYWORD_ANALYZER;
+
+/**
+ * A special field mapper for multi-valued keywords that may contain duplicate values. If the associated <code>counted_terms</code>
+ * aggregation is used, duplicates are considered in aggregation results. Consider the following values:
+ *
+ * <ul>
+ *     <li><code>["a", "a", "b"]</code></li>
+ *     <li><code>["a", "b", "b"]</code></li>
+ * </ul>
+ *
+ * While a regular <code>keyword</code> and the corresponding <code>terms</code> aggregation deduplicates values and reports a count of
+ * 2 for each key (one per document), a <code>counted_terms</code> aggregation on a <code>counted_keyword</code> field will consider
+ * the actual count and report a count of 3 for each key.
+ */
+public class CountedKeywordFieldMapper extends FieldMapper {
```

Comment:
I don't this new field type now support synthetic source (the `syntheticFieldLoader()` isn't overwritten).
Is the plan to use this any time soon? If not, then it is ok to keep it the way it is. However we should then avoid this field from being used in other use cases.

Reply:
For the time being we don't support synthetic source and also have no intention to do so. Therefore I think it's ok to keep this as is.

Reply:
Sounds good. Can you add this as a comment to the java docs?